File: writer/algorithm/max-sum-of-subsequence/maxsum.scan.c

/********************************************************************
created:    2017-04-20
author:     lixianmin#live.cn

Copyright (C) - All Rights Reserved
*********************************************************************/

#include <assert.h>
#include <stdio.h>
#include <stdlib.h>

// 这个算法在数组值全负的情况下会返回最大的一个负数
int maxSum (int data[], int length)
{
    if (NULL == data || length <= 0)
    {
        return 0;
    }

    int maxSoFar = data[0];
    int maxEndingHere = 0;
    for (int i= 1; i< length; ++i)
    {
        maxEndingHere += data[i];
        if (maxEndingHere > maxSoFar)
        {
            maxSoFar = maxEndingHere;
        }
        else if (maxEndingHere < 0)
        {
            maxEndingHere = 0;
        }
    }

    return maxSoFar;
}

void test (int data[], int length)
{
    int max = maxSum (data, length);
    printf("max=%d\n", max);
}

int main ()
{
    int data[] = {31, -41, 59, 26, -53, 58, 97, -93, -23, 84};
    test (data, sizeof(data)/sizeof(data[0]));
    return 0;
}

File: writer/algorithm/maximum-subarray/maxsum.scan.rb
#!/usr/bin/ruby
#encoding:utf-8
#
######################################################################
# created:    2017-04-20
# author:     lixianmin#live.cn
#
# Copyright (C) - All Rights Reserved
######################################################################

# 这个算法在数组值全负的情况下会返回最大的一个负数
def max_sum (data)
    return if !data or data.length <= 0

    maxSoFar, maxEndingHere = 0, 0
    data.each do |item|
        maxEndingHere += item
        if maxEndingHere > maxSoFar
            maxSoFar = maxEndingHere
        elsif maxEndingHere < 0
            maxEndingHere = 0
        end
    end

    maxSoFar;
end

def test (data)
    max = max_sum (data)
    puts "[#{data.join(',')}] => #{max}"
end

def main
    test [1, 3, -2, 4, -5]
    test [31, -41, 59, 26, -53, 58, 97, -93, -23, 84]
end

if __FILE__ == $0
    main
end

